crates/oxc_linter/src/rules/promise/no_return_wrap.rs

```diff
@@ -0,0 +1,299 @@
+    ///   return val * 2
+    ///   })
+    /// myPromise.then(function (val) {
+    ///     throw 'bad thing'
+    /// })
+    /// ```
+    NoReturnWrap,
+    style,
+);
+
+impl Rule for NoReturnWrap {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let config = value.get(0);
+
+        let allow_reject = config
+            .and_then(|c| c.get("allowReject"))
+            .and_then(serde_json::Value::as_bool)
+            .map_or(false, |v| v);
+
+        Self(Box::new(NoReturnWrapOptions { allow_reject }))
+    }
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::CallExpression(call_expr) = node.kind() else {
+            return;
+        };
+        if call_expr.callee.as_member_expression().is_some()
+            && is_promise(call_expr).is_some()
+            && is_in_promise(node.id(), ctx)
+        {
+            if matches!(call_expr.callee_name(), Some("resolve")) {
+                ctx.diagnostic(no_return_wrap_diagnostic_resolve(call_expr.span));
+            }
+            if !self.0.allow_reject && matches!(call_expr.callee_name(), Some("reject")) {
+                ctx.diagnostic(no_return_wrap_diagnostic_reject(call_expr.span));
+            }
+        }
+    }
+}
+
+fn is_in_promise(id: NodeId, ctx: &LintContext) -> bool {
+    // find the innermost function node
+    let mut function_node = ctx.nodes().ancestors(id).find(|node| {
+        matches!(node.kind(), AstKind::ArrowFunctionExpression(_) | oxc_ast::AstKind::Function(_))
+    });
+
+    // move up to the parent node while the function is re-bound with "bind"
+    while let Some(func_node) = function_node {
+        let Some(mut parent) = ctx.nodes().parent_id((func_node).id()) else { break };
+        if matches!(ctx.nodes().kind(parent), ParenthesizedExpression(_)) {
+            parent = match ctx.nodes().parent_id(parent) {
+                Some(id) => id,
+                None => break,
+            };
+        }
+        let oxc_ast::AstKind::MemberExpression(member_expr) = ctx.nodes().kind(parent) else {
+            break;
+        };
+        if member_expr.static_property_name() != Some("bind") {
+            break;
+        };
+        let Some(grandparent) = ctx.nodes().parent_id(parent) else { break };
+        let oxc_ast::AstKind::CallExpression(call_expr) = ctx.nodes().kind(grandparent) else {
+            break;
+        };
+        if call_expr.callee.as_member_expression().is_some() {
+            function_node = Some(ctx.nodes().get_node(grandparent));
+            continue;
+        }
+
+        break;
+    }
+    let Some(function_node) = function_node else {
+        return false;
+    };
+
+    // judge parent node is promise
+    let Some(parent) = ctx.nodes().parent_node(function_node.id()) else { return false };
+    let parent_kind = parent.kind();
+    let parent = match parent_kind {
+        AstKind::ParenthesizedExpression(_) | AstKind::Argument(_) => parent,
```

Comment:
Some patterns have different AST node depths compared to espree, so I move to the parent node. There might be a non-ad-hoc solution for this.